src/System.Net.Requests/src/System/Net/FtpWebRequest.cs

```diff
@@ -0,0 +1,2544 @@
+                            {
+                                if (m_RequestStage < RequestStage.ReadReady)
+                                    m_ReadAsyncResult = new LazyAsyncResult(null, null, null);
+                            }
+
+                            // GetRequeststream or BeginGetRequestStream has not finished yet?
+                            if (m_ReadAsyncResult != null)
+                                m_ReadAsyncResult.InternalWaitForCompletion();
+
+                            CheckError();
+                        }
+                    }
+                    else
+                    {
+#if FTP_FEATURE_CACHE
+                        do
+                        {
+#endif
+                            SubmitRequest(false);
+                            if (m_MethodInfo.IsUpload)
+                                FinishRequestStage(RequestStage.WriteReady);
+                            else
+                                FinishRequestStage(RequestStage.ReadReady);
+                            CheckError();
+#if FTP_FEATURE_CACHE
+                        } while (!CheckCacheRetrieveOnResponse());
+#endif
+
+                        EnsureFtpWebResponse(null);
+
+#if FTP_FEATURE_CACHE
+                        // This may update the Stream memeber on m_FtpWebResponse based on the CacheProtocol feedback.
+                        CheckCacheUpdateOnResponse();
+
+                        if (m_FtpWebResponse.IsFromCache)
+                            FinishRequestStage(RequestStage.ReleaseConnection);
+#endif
+                    }
+                }
+
+                statusCode = GetStatusCode(m_FtpWebResponse);
+#if FEATURE_FTP_TRACING
+                success = true;
+#endif
+            }
+            catch (Exception exception) {
+#if FEATURE_FTP_TRACING
+                if (FrameworkEventSource.Log.IsEnabled()) {
+                    WebException webException = exception as WebException;
+                    if (webException != null) {
+                        statusCode = GetStatusCode(webException);        
+                    }
+                }
+#endif
+
+#if FEATURE_FTP_LOGGING
+                if(Logging.On)Logging.Exception(Logging.Web, this, "GetResponse", exception);
+#endif
+
+                // if m_Exception == null, we are about to throw an exception to the user
+                // and we haven't saved the exception, which also means we haven't dealt
+                // with it. So just release the connection and log this for investigation
+                if (m_Exception == null) {
+#if FEATURE_FTP_LOGGING
+                    if(Logging.On)Logging.PrintWarning(Logging.Web, SR.GetString(SR.net_log_unexpected_exception, "GetResponse()"));
+#endif
+                    if (!ExceptionCheck.IsFatal(exception)){
+                        if (GlobalLog.IsEnabled)
+                        {
+                            GlobalLog.Assert("Find out why we are getting an unexpected exception.");
```

Comment:
/cc @stephentoub 
Changes in behavior of GlobalLog.Assert* implementation requires adding Debug.Fail outside the check (see examples in System.Net.Sockets).